power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -21,114 +25,242 @@ namespace detail = power_grid_model::optimizer::detail;
+                                                    (tap_at_first_side || transformer3w.tap_side() == second_side);
+            if (single_direction_condition) {
+                auto const& tap_side_node = tap_at_first_side ? from_node : to_node;
+                auto const& non_tap_side_node = tap_at_first_side ? to_node : from_node;
+                Idx2D const regulated_idx =
+                    assigned_idx ? unregulated_idx : main_core::get_component_idx_by_id(state, transformer3w.id());
+                assigned_idx = true;
+                add_to_edge(edges, edge_props, tap_side_node, non_tap_side_node, {regulated_idx, 1});
+            } else {
+                add_to_edge(edges, edge_props, from_node, to_node, {unregulated_idx, 1});
+                add_to_edge(edges, edge_props, to_node, from_node, {unregulated_idx, 1});
+            }
+        }
+    }
+}
+
+template <std::derived_from<Transformer> Component, class ComponentContainer>
+    requires main_core::model_component_state_c<main_core::MainModelState, ComponentContainer, Component>
+constexpr void add_edge(main_core::MainModelState<ComponentContainer> const& state,
+                        RegulatedObjects const& regulated_objects, TrafoGraphEdges& edges,
+                        TrafoGraphEdgeProperties& edge_props) {
+    for (auto const& transformer : state.components.template citer<Transformer>()) {
+        if (!transformer.from_status() || !transformer.to_status()) {
+            continue;
+        }
+        auto const& from_node = transformer.from_node();
+        auto const& to_node = transformer.to_node();
+
+        if (regulated_objects.transformers.contains(transformer.id())) {
+            auto const tap_at_from_side = transformer.tap_side() == BranchSide::from;
+            auto const& tap_side_node = tap_at_from_side ? from_node : to_node;
+            auto const& non_tap_side_node = tap_at_from_side ? to_node : from_node;
+            if (get_component<Node>(state, tap_side_node).u_rated() <
+                get_component<Node>(state, non_tap_side_node).u_rated()) {
+                throw AutomaticTapCalculationError(transformer.id());
+            }
+            add_to_edge(edges, edge_props, tap_side_node, non_tap_side_node,
+                        {main_core::get_component_idx_by_id(state, transformer.id()), 1});
+        } else {
+            add_to_edge(edges, edge_props, from_node, to_node, {unregulated_idx, 1});
+            add_to_edge(edges, edge_props, to_node, from_node, {unregulated_idx, 1});
+        }
+    }
+}
+
+template <typename Component>
+concept non_regulating_branch_c = std::derived_from<Link, Component> || std::derived_from<Line, Component>;
+
+template <non_regulating_branch_c Component, class ComponentContainer>
+    requires main_core::model_component_state_c<main_core::MainModelState, ComponentContainer, Component>
```

Comment:
more general:

```suggestion
template <typename Component>
concept non_regulated_branch_c = std::derived_from<Component, Branch> && !transformer_c<Component>;

template <non_regulating_branch_c Component, class ComponentContainer>
    requires main_core::model_component_state_c<main_core::MainModelState, ComponentContainer, Component>
```

since you only use it once, you can also keep this logic but make it as follows:

```suggestion
template <std::derived_from<Component, Branch> Component, class ComponentContainer>
    requires main_core::model_component_state_c<main_core::MainModelState, ComponentContainer, Component> && !transformer_c<Component>
```

because then the compiler will raise an error if the edge type is not derived from `Branch` and also not a `Transformer` or `ThreeWindingTransformer`. The following 2 such cases may arise:

* For edge-type components that are not `Branch`, the user should specify overloads to create the unregulated edges.
* For transformer types (that can be regulated) that are not `Transformer` or `ThreeWindingTransformer`, the user should also specify overloads